File: nobodywho/python/examples/encoder_demo.py
#!/usr/bin/env python3
"""
Simple embeddings demo showing how to generate embeddings and compute similarity.
Usage: python embeddings_demo.py /path/to/embeddings/model.gguf
"""

import sys
import nobodywho


def main():
    if len(sys.argv) != 2:
        print("Usage: python embeddings_demo.py /path/to/embeddings/model.gguf")
        sys.exit(1)

    model_path = sys.argv[1]
    print(f"Loading embeddings model: {model_path}")

    # Create model and encoder instance
    model = nobodywho.Model(model_path, use_gpu_if_available=True)
    encoder = nobodywho.Encoder(model, n_ctx=2048)

    # Example texts
    texts = [
        "The quick brown fox jumps over the lazy dog.",
        "A fast brown fox leaps over the sleepy dog.",
        "The weather is sunny today.",
        "Machine learning is a subset of artificial intelligence.",
        "Deep learning uses neural networks with many layers.",
    ]

    print("\nGenerating embeddings...")
    text_embeddings = []
    for i, text in enumerate(texts):
        print(f"  {i + 1}. {text}")
        embedding = encoder.encode(text)
        text_embeddings.append((text, embedding))

    print(f"\nEmbedding dimension: {len(text_embeddings[0][1])}")

    # Compute pairwise similarities
    print("\nPairwise cosine similarities:")
    for i in range(len(texts)):
        for j in range(i + 1, len(texts)):
            similarity = nobodywho.cosine_similarity(
                text_embeddings[i][1], text_embeddings[j][1]
            )
            print(f"  Text {i + 1} vs Text {j + 1}: {similarity:.3f}")

    print("\nDone!")


if __name__ == "__main__":
    main()

File: nobodywho/js/examples/encoder_demo.mjs
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { join, dirname } from 'node:path';

const pkgDir = join(dirname(fileURLToPath(import.meta.url)), '..', 'pkg-bundler');
const { default: createNobodyWhoModule } = await import(join(pkgDir, 'nobodywho_js.js'));
const m = await createNobodyWhoModule({ locateFile: (p) => join(pkgDir, p) });
m.init();

const model = await m.Model.loadBytes(new Uint8Array(readFileSync(process.argv[2])));
const encoder = new m.Encoder(model, 2048);

const texts = [
  'The quick brown fox jumps over the lazy dog.',
  'A fast brown fox leaps over the sleepy dog.',
  'The weather is sunny today.',
  'Machine learning is a subset of artificial intelligence.',
  'Deep learning uses neural networks with many layers.',
];

const embeddings = [];
for (const text of texts) embeddings.push(await encoder.encode(text));

const cosine = (a, b) => {
  let dot = 0, na = 0, nb = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  return dot / Math.sqrt(na * nb);
};

for (let i = 0; i < texts.length; i++) {
  for (let j = i + 1; j < texts.length; j++) {
    console.log(`${i + 1} vs ${j + 1}: ${cosine(embeddings[i], embeddings[j]).toFixed(3)}`);
  }
}
